### PR TITLE
PR to set SNI during boost lib websocket next_layer() SSL handshake process happen during launch of app tunneling.

### PR DESCRIPTION
**Pull request by jaiswal-sumit**
This PR to set SNI during boost lib websocket next_layer() SSL handshake process happen during launch of app tunneling. 

### Motivation
- We have a use case for running AWS IoT secure tunnel within a Customer Managed VPN, along with a proxy service infrastructure to route multiple external services on a single domain. During the Customer Managed VPN setup, we noticed that the SSL handshake was failing when launching the local proxy. After a detailed investigation of the issue, we found that the code was not setting the Server Name Indication (SNI) during the initial SSL handshake call in the 'Client Hello' of the next_layer() SSL handshake code.
- This issue is linked to the Boost Beast WebSocket Secure (WSS) stream library code, which does not supply SNI by default during the next_layer() async_ssl_handshake() function. Therefore, following the example provided by the Boost library, it is suggested to use the SSL_set_tlsext_host_name function to set the SNI.
- This PR aims to resolve the SSL handshake issue and ensure that the code functions correctly in both non-VPN and VPN application tunnels.
- Issue number: AWS Case id: 13813696371

#### Change summary
Supplied host string From TcpAdapter which initiates async_ssl_hanshake
Extend WebSocketStream async_ssl_hanshake to take host string and use that to set SNI.
WebSocketSteam async_ssl_hanshake() code added with SSL_set_tlsext_host_name to set SNI with host string.

#### Revision diff summary
NA

### Testing
 **Is your change tested? Yes
 **Please list your testing steps and test results.** 
Login into CMVPN infra, 
Deploy the fixed code localproxy execuable 
open the tunnel. 
Check the connection status 
Connection status found connected at both sides
- CI test run result: NA

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by RogerZhongAWS** [comment]
Hey, thanks the description of your issue and for your contribution. This looks fine to me, will merge as soon as we get a second approval on this.

**Review by RogerZhongAWS — APPROVED** [review]

File: src/WebSocketStream.cpp
Comment (kareali) [inline src/WebSocketStream.cpp:197]: Since this block is executed when the web proxy is not using TLS, Is it needed to set the SNI in this code branch/path?

File: src/WebSocketStream.h
Comment (kareali) [inline src/WebSocketStream.h:153]: Could you please rename it to `web_proxy_host` and also add it as a param on line 150? We used `host` [here](https://github.com/aws-samples/aws-iot-securetunneling-localproxy/blob/b6c31b442fa6d4c930e705d2853efd73dadff169/src/WebSocketStream.h#L164C22-L164C40) to refer the websocket host/endpoint.

**Review by kareali — CHANGES_REQUESTED** [review]

File: src/WebSocketStream.cpp
Comment (HarshGandhi-AWS) [reply src/WebSocketStream.cpp:197]: Hey @jaiswal-sumit , can you please address this and other comments raised by the team? Thank you

Comment (jaiswal-sumit) [reply src/WebSocketStream.cpp:197]: Hello @HarshGandhi-AWS @kareali Its fine to set SNI even during web proxy connection calls too.
Below are the debug logs: As example: I have use regional aws default tunnel, in default behaviour with web proxy setting. Setting SNI is need in actual request which even go via Web proxy as server needs SNI information. I am open to discuss more on it, if it's needed. Thank you :)  

```
[2023-10-11 08:20:20.541774] (0x00007ff849dfd700) [trace] Establishing TCP connection with the Web Proxy
[2023-10-11 08:20:20.542750] (0x00007ff849dfd700) [debug] Connected successfully with Web Proxy
[2023-10-11 08:20:20.542771] (0x00007ff849dfd700) [trace] Calling lowest_layer with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:20.542789] (0x00007ff849dfd700) [trace] Preparing HTTP CONNECT request
[2023-10-11 08:20:20.542800] (0x00007ff849dfd700) [trace] Sending HTTP CONNECT
[2023-10-11 08:20:20.542958] (0x00007ff849dfd700) [debug] Successfully sent HTTP CONNECT to the Web proxy
[2023-10-11 08:20:20.542990] (0x00007ff849dfd700) [trace] Waiting for HTTP CONNECT response from the Web proxy
[2023-10-11 08:20:20.811438] (0x00007ff849dfd700) [trace] Parsing the HTTPS response from the Web proxy

[2023-10-11 08:20:20.832312] (0x00007ff849dfd700) [info] TCP tunnel established successfully
[2023-10-11 08:20:20.832341] (0x00007ff849dfd700) [debug] Connected successfully with proxy server
[2023-10-11 08:20:20.832522] (0x00007ff849dfd700) [trace] Calling lowest_layer with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:20.832544] (0x00007ff849dfd700) [trace] Calling lowest_layer with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:20.832555] (0x00007ff849dfd700) [trace] Performing SSL handshake with proxy server
[2023-10-11 08:20:20.832565] (0x00007ff849dfd700) [trace] Calling set_verify_mode with type: single_ssl_stream
[2023-10-11 08:20:20.832574] (0x00007ff849dfd700) [trace] Calling set_verify_callback with type: single_ssl_stream
[2023-10-11 08:20:20.832586] (0x00007ff849dfd700) [trace] Calling next_layer().async_handshake with type: websocket_stream_single_ssl_type
**[2023-10-11 08:20:20.832596] (0x00007ff849dfd700) [trace] SSL next_layer() SNI is set : data.tunneling.iot.eu-west-1.amazonaws.com**
[2023-10-11 08:20:21.219403] (0x00007ff849dfd700) [debug] Successfully completed SSL handshake with proxy server
[2023-10-11 08:20:21.219451] (0x00007ff849dfd700) [trace] Performing websocket handshake with proxy server
[2023-10-11 08:20:21.219473] (0x00007ff849dfd700) [trace] Calling async_handshake with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:21.219536] (0x00007ff849dfd700) [trace] Web socket ugprade request(*not entirely final):


[2023-10-11 08:20:21.443448] (0x00007ff849dfd700) [debug] Web socket subprotocol selected: aws.iot.securetunneling-3.0
[2023-10-11 08:20:21.443465] (0x00007ff849dfd700) [info] Successfully established websocket connection with proxy server: wss://data.tunneling.iot.eu-west-1.amazonaws.com:443
```

**Review by abeytr — APPROVED** [review]
We tested this for the following workflows and it is working as expected:
1. localProxy connection over private urls
2. localProxy connection over public urls

**Review by kareali — APPROVED** [review]

**Review by RogerZhongAWS — APPROVED** [review]

**Pull request merged**
